src/python/pants/engine/scheduler.py

```diff
@@ -285,36 +285,41 @@ def _run_and_return_roots(self, session, execution_request):
       self._native.lib.nodes_destroy(raw_roots)
     return roots
 
-  def capture_snapshots(self, path_globs_and_roots):
+  def capture_snapshots(self, session, path_globs_and_roots):
     """Synchronously captures Snapshots for each matching PathGlobs rooted at a its root directory.
 
     This is a blocking operation, and should be avoided where possible.
 
+    :param session Session: the session in which work units shall be recorded.
     :param path_globs_and_roots tuple<PathGlobsAndRoot>: The PathGlobs to capture, and the root
            directory relative to which each should be captured.
     :returns: A tuple of Snapshots.
     """
     result = self._native.lib.capture_snapshots(
       self._scheduler,
+      session,
       self._to_value(_PathGlobsAndRootCollection(path_globs_and_roots)),
     )
     return self._raise_or_return(result)
 
-  def merge_directories(self, directory_digests):
+  def merge_directories(self, session, directory_digests):
     """Merges any number of directories.
 
+    :param session Session: the session in which work units shall be recorded.
     :param directory_digests: Tuple of DirectoryDigests.
     :return: A Digest.
     """
     result = self._native.lib.merge_directories(
       self._scheduler,
+      session,
       self._to_value(_DirectoryDigests(directory_digests)),
     )
     return self._raise_or_return(result)
 
-  def materialize_directories(self, directories_paths_and_digests):
+  def materialize_directories(self, session, directories_paths_and_digests):
```

Comment:
Pierre, thank you! 
I have a minor question comment.
`SchedulerSession` (Session) has a scheduler. As we pass to the function `materialize_directories`  the scheduler but also the session, maybe there is no need for function `materialize_directories` for the scheduler and the session and it can be a `SchedulerSession` function?

Reply:
Thanks for your review. I see your point. I can merge both functions (same for `merge_directories` and `capture_snapshots`). It does add a few "trainwrecks", but it removes one level of indirection. probably a good thing overall.

Reply:
Done in the last 3 commits